core/src/avm2/object/class_object.rs

```diff
@@ -304,75 +308,70 @@ impl<'gc> ClassObject<'gc> {
         let class = write.class;
         let scope = write.class_scope;
 
-        let interface_names = class.read().interfaces().to_vec();
+        let interface_names = class.read().direct_interfaces().to_vec();
         let mut interfaces = Vec::with_capacity(interface_names.len());
-        for interface_name in interface_names {
-            interfaces.push(self.early_resolve_interface(scope.domain(), &interface_name)?);
-        }
 
-        if !interfaces.is_empty() {
-            write.interfaces = interfaces;
+        let mut dedup = HashSet::new();
+        let mut queue = vec![class];
+        while let Some(cls) = queue.pop() {
+            for interface_name in cls.read().direct_interfaces() {
+                let interface = self.early_resolve_class(scope.domain(), interface_name)?;
+
+                if !interface.read().is_interface() {
+                    return Err(format!(
+                        "Class {:?} is not an interface and cannot be implemented by classes",
+                        interface.read().name().local_name()
+                    )
+                    .into());
+                }
+
+                if dedup.insert(ClassHashWrapper(interface)) {
+                    queue.push(interface);
+                    interfaces.push(interface);
+                }
+            }
+
+            if let Some(superclass_name) = cls.read().super_class_name() {
+                queue.push(self.early_resolve_class(scope.domain(), superclass_name)?);
+            }
         }
+        write.interfaces = interfaces.clone();
 
         //At this point, we need to reresolve *all* interface traits.
         //Otherwise we won't get overrides.
         drop(write);
 
-        let mut class = Some(self);
-
-        while let Some(cls) = class {
-            let mut interfaces = cls.interfaces();
-            while let Some(interface) = interfaces.pop() {
-                let iface_read = interface.read();
-                for interface_trait in iface_read.instance_traits() {
-                    if !interface_trait.name().namespace().is_public() {
-                        let public_name = QName::new(
-                            activation.context.avm2.public_namespace,
-                            interface_trait.name().local_name(),
-                        );
-                        self.instance_vtable().copy_property_for_interface(
-                            activation.context.gc_context,
-                            public_name,
-                            interface_trait.name(),
-                        );
-                    }
+        for interface in interfaces {
```

Comment:
note: avmplus only does it for newly implemented interfaces (as the old ones are already properly copied in the vtable).
That said, I can't tell for sure if this can be observed. They _do_ have some slightly fancier logic that limits when interface namespace is copied; I can try look for some edge cases later.

Reply:
AVM2 code can only observe interfaces through `is/instanceof` and property lookups, right?